src/cargo/ops/registry.rs

```diff
@@ -943,18 +944,28 @@ fn get_source_id(
             bail!("crates-io is replaced with non-remote-registry source {replacement_sid};\ninclude `--registry crates-io` to use crates.io");
         }
     } else {
-        Ok((builtin_replacement_sid, sid))
+        Ok(PublishSourceIds {
+            original: sid,
+            replacement: builtin_replacement_sid,
+        })
     }
 }
 
+struct PublishSourceIds {
```

Comment:
I'm not sure `PublishSourceIds` makes sense, since it's used for API operations other than publish.
`RegistrySourceIds`?

May want to add another block comment further explaining what the replaced source is. Something like:

> The source for `crates.io` may be replaced by a built-in source for accessing crates.io with the sparse protocol, or a source for the testing framework (when the `replace_crates_io` function is used). User-defined source replacement is not applied.

Reply:
Thanks!